dbms/src/IO/Compression/EncodingUtil.h

```diff
@@ -0,0 +1,271 @@
+    BitpackingPrimitives::packBuffer(reinterpret_cast<unsigned char *>(dest), values.data(), values.size(), width);
+    return sizeof(T) + sizeof(UInt8) + required_size;
+}
+
+template <std::integral T>
+void applyFrameOfReference(T * dst, T frame_of_reference, UInt32 count);
+
+template <std::integral T>
+void FORDecoding(const char * src, UInt32 source_size, char * dest, UInt32 dest_size)
+{
+    UInt8 bytes_size = sizeof(T);
+    if unlikely (dest_size % bytes_size != 0)
+        throw Exception(
+            ErrorCodes::CANNOT_DECOMPRESS,
+            "uncompressed size {} is not aligned to {}",
+            dest_size,
+            bytes_size);
+    const auto count = dest_size / sizeof(T);
+    T frame_of_reference = unalignedLoad<T>(src);
+    src += sizeof(T);
+    auto width = unalignedLoad<UInt8>(src);
+    src += sizeof(UInt8);
+    const auto required_size = source_size - sizeof(T) - sizeof(UInt8);
+    RUNTIME_CHECK(BitpackingPrimitives::getRequiredSize(count, width) == required_size);
+    auto round_size = BitpackingPrimitives::roundUpToAlgorithmGroupSize(count);
+    if (round_size != count)
+    {
+        // Reserve enough space for the temporary buffer.
+        unsigned char tmp_buffer[round_size * sizeof(T)];
+        BitpackingPrimitives::unPackBuffer<T>(tmp_buffer, reinterpret_cast<const unsigned char *>(src), count, width);
+        applyFrameOfReference(reinterpret_cast<T *>(tmp_buffer), frame_of_reference, count);
+        memcpy(dest, tmp_buffer, dest_size);
+        return;
+    }
+    BitpackingPrimitives::unPackBuffer<T>(
+        reinterpret_cast<unsigned char *>(dest),
+        reinterpret_cast<const unsigned char *>(src),
+        count,
+        width);
+    applyFrameOfReference(reinterpret_cast<T *>(dest), frame_of_reference, count);
+}
+
+/// Delta encoding
+
+template <std::integral T>
+void deltaEncoding(const T * source, UInt32 count, T * dest)
+{
+    T prev = 0;
+    for (UInt32 i = 0; i < count; ++i)
+    {
+        T curr = source[i];
+        dest[i] = curr - prev;
+        prev = curr;
+    }
+}
+
+template <std::integral T>
+void ordinaryDeltaDecoding(const char * source, UInt32 source_size, char * dest)
+{
+    T accumulator{};
+    const char * const source_end = source + source_size;
+    while (source < source_end)
+    {
+        accumulator += unalignedLoad<T>(source);
+        unalignedStore<T>(dest, accumulator);
+
+        source += sizeof(T);
+        dest += sizeof(T);
+    }
+}
+
+template <std::integral T>
+void deltaDecoding(const char * source, UInt32 source_size, char * dest);
+
+/// Delta + Frame of Reference encoding
+
+template <std::integral T>
+void ordinaryDeltaFORDecoding(const char * src, UInt32 source_size, char * dest, UInt32 dest_size)
+{
+    using TS = typename std::make_signed_t<T>;
+    // copy first value to dest
+    memcpy(dest, src, sizeof(T));
+    if (unlikely(source_size <= sizeof(T)))
+        return;
```

Comment:
The check of `source_size > sizeof(T)` should be done before `memcpy`?

Reply:
`==` here, fixed.

The following make sure `source_size >= sizeof(T)`.

```cpp
void CompressionCodecDeltaFOR::doDecompressData(
    const char * source,
    UInt32 source_size,
    char * dest,
    UInt32 uncompressed_size) const
{
    if (unlikely(source_size < 2))
        throw Exception(
            ErrorCodes::CANNOT_DECOMPRESS,
            "Cannot decompress DeltaFor-encoded data. File has wrong header");
```